apps/emqx_bridge_snowflake/src/emqx_bridge_snowflake_connector.erl

```diff
@@ -0,0 +1,959 @@
+        max_retries := MaxRetries
+    } = HTTPClientConfig,
+    JWTToken = emqx_connector_jwt:ensure_jwt(JWTConfig),
+    AuthnHeader = [<<"BEARER ">>, JWTToken],
+    Headers = http_headers(AuthnHeader),
+    Body = emqx_utils_json:encode(#{files => StagedFiles}),
+    %% TODO: generate unique request id
+    Req = {InserFilesPath, Headers, Body},
+    ?tp(debug, "snowflake_stage_insert_files_request", #{
+        action_res_id => HTTPPool,
+        staged_files => StagedFiles
+    }),
+    ?MODULE:do_insert_files_request(HTTPPool, Req, RequestTTL, MaxRetries).
+
+%% Exposed for mocking
+do_insert_files_request(HTTPPool, Req, RequestTTL, MaxRetries) ->
+    ehttpc:request(HTTPPool, post, Req, RequestTTL, MaxRetries).
+
+insert_report_request(HTTPPool, Opts, HTTPClientConfig) ->
+    #{
+        jwt_config := JWTConfig,
+        insert_report_path := InsertReportPath0,
+        request_ttl := RequestTTL,
+        max_retries := MaxRetries
+    } = HTTPClientConfig,
+    JWTToken = emqx_connector_jwt:ensure_jwt(JWTConfig),
+    AuthnHeader = [<<"BEARER ">>, JWTToken],
+    Headers = http_headers(AuthnHeader),
+    InsertReportPath =
+        case Opts of
+            #{begin_mark := BeginMark} when is_binary(BeginMark) ->
+                <<InsertReportPath0/binary, "?beginMark=", BeginMark/binary>>;
+            _ ->
+                InsertReportPath0
+        end,
+    Req = {InsertReportPath, Headers},
+    Response = ehttpc:request(
+        HTTPPool,
+        get,
+        Req,
+        RequestTTL,
+        MaxRetries
+    ),
+    case Response of
+        {ok, 200, _Headers, Body0} ->
+            Body = emqx_utils_json:decode(Body0, [return_maps]),
+            {ok, Body};
+        _ ->
+            {error, Response}
+    end.
+
+http_headers(AuthnHeader) ->
+    [
+        {<<"X-Snowflake-Authorization-Token-Type">>, <<"KEYPAIR_JWT">>},
+        {<<"Content-Type">>, <<"application/json">>},
+        {<<"Authorization">>, AuthnHeader}
+    ].
+
+row_to_map(Row0, Headers) ->
+    Row1 = tuple_to_list(Row0),
+    Row2 = lists:map(fun emqx_utils_conv:bin/1, Row1),
+    Row = lists:zip(Headers, Row2),
+    maps:from_list(Row).
+
+action_status(ActionResId, #{mode := aggregated} = ActionState) ->
+    #{
+        aggreg_id := AggregId,
+        http := #{connect_timeout := ConnectTimeout}
+    } = ActionState,
+    %% NOTE: This will effectively trigger uploads of buffers yet to be uploaded.
+    Timestamp = erlang:system_time(second),
+    ok = emqx_connector_aggregator:tick(AggregId, Timestamp),
+    case http_pool_workers_healthy(ActionResId, ConnectTimeout) of
+        true ->
+            ?status_connected;
+        false ->
+            ?status_disconnected
+    end.
+
+stage_file_sql(Filename, Database, Schema, Stage) ->
+    SQL0 = iolist_to_binary([
+        <<"PUT file://">>,
+        %% TODO: use action as directory name on stage?
+        Filename,
+        <<" @">>,
+        %% TODO: escape names?
```

Comment:
plan to resolve it ?

Reply:
have yet to figure out the proper way to escape things safely, specially if someone decides to use double quotes, back ticks, etc. :see_no_evil:

Reply:
can maybe directly return error for such strings.

Reply:
I'll see if it works well with `odbc:param_query`

Reply:
indeed, `PUT` doesn't seem to support bind variables :thinking:

Reply:
added escaping for the identifiers

:checkered_flag: